components/supervisor/pkg/ports/ports.go

```diff
@@ -0,0 +1,575 @@
+		localPort := served.Port
+		_, exists := pm.proxies[localPort]
+		if exists || !served.BoundToLocalhost {
+			continue
+		}
+
+		var globalPort uint32
+		for port := proxyPortRangeHi; port >= proxyPortRangeLo; port-- {
+			if _, used := opened[port]; used {
+				continue
+			}
+			if _, used := pm.internal[port]; used {
+				continue
+			}
+
+			globalPort = port
+			break
+		}
+		if globalPort == 0 {
+			log.WithField("port", localPort).Error("cannot find a free proxy port")
+			continue
+		}
+
+		proxy, err := pm.proxyStarter(localPort, globalPort)
+		if err != nil {
+			log.WithError(err).WithField("globalPort", globalPort).WithField("localPort", localPort).Warn("cannot start localhost proxy")
+			continue
+		}
+		log.WithField("globalPort", globalPort).WithField("localPort", localPort).Info("localhost proxy has been started")
+
+		pm.internal[globalPort] = struct{}{}
+		pm.proxies[localPort] = &localhostProxy{
+			Closer:    proxy,
+			proxyPort: globalPort,
+		}
+	}
+}
+
+func (pm *Manager) updateState() {
+	var added, updated, removed []uint32
+	defer func() {
+		pm.publishStatus(added, updated, removed)
+	}()
+
+	newState := pm.nextState()
+	for port := range newState {
+		_, exists := pm.state[port]
+		if !exists {
+			added = append(added, port)
+		}
+	}
+	for port, mp := range pm.state {
+		newMp, exists := newState[port]
+		if !exists {
+			removed = append(removed, port)
+			continue
+		}
+		if cmp.Equal(newMp, mp) {
+			continue
+		}
+		updated = append(updated, port)
+	}
+	pm.state = newState
+}
+
+func (pm *Manager) nextState() map[uint32]*managedPort {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	state := make(map[uint32]*managedPort)
+
+	// 1. first capture exposed since they don't depend on configured or served ports
+	for _, exposed := range pm.exposed {
+		port := exposed.LocalPort
+		if pm.boundInternally(port) {
+			continue
+		}
+
+		config, exists := pm.configs.Get(port)
+		if !exists {
+			config, _ = pm.configs.GetFromRange(port)
+		}
+		state[port] = &managedPort{
+			LocalhostPort: port,
+			GlobalPort:    exposed.GlobalPort,
+			Exposed:       true,
+			Public:        exposed.Public,
+			URL:           exposed.URL,
+			OnExposed:     getOnExposed(config, port),
+		}
+	}
+
+	// 2. second capture configured since we don't want to auto expose already exposed ports
+	if pm.configs != nil {
+		pm.configs.ForEach(func(port uint32, config *gitpod.PortConfig) {
+			if pm.boundInternally(port) {
+				return
+			}
+
+			mp, exists := state[port]
+			if !exists {
+				mp = &managedPort{}
+				state[port] = mp
+			}
+			mp.LocalhostPort = port
+
+			if mp.Exposed {
+				return
+			}
+			mp.OnExposed = getOnExposed(config, port)
+			mp.Public = config.Visibility != "private"
+			err := pm.E.Expose(ctx, mp.LocalhostPort, mp.GlobalPort, mp.Public)
+			if err != nil {
+				log.WithError(err).WithField("port", *mp).Warn("cannot auto-expose port")
+				return
```

Comment:
```suggestion
continue
```

Just because one port failed we might not want to fail the others

Reply:
that's not for loop, but a callback function for each port config